src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+        let! ct = Async.CancellationToken
+        let! t, slice = readSliceAsync conn streamName batchSize startPos ct |> Async.AwaitTaskCorrect |> Stopwatch.Time
+        let bytes, count = slice.Messages |> Array.sumBy resolvedEventLen, slice.Messages.Length
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let evt = Log.Slice reqMetric
+        let log = if not (log.IsEnabled Events.LogEventLevel.Debug) then log else log |> Log.propResolvedEvents "Json" slice.Messages
+        (log |> Log.prop "startPos" startPos |> Log.prop "bytes" bytes |> Log.event evt).Information("Mdb{action:l} count={count} version={version}",
+            "Read", count, slice.LastVersion)
+        return slice }
+
+    let private readBatches (log : ILogger) (readSlice : int64 -> ILogger -> Async<StreamEventsSlice>)
+            (maxPermittedBatchReads : int option) (startPosition : int64)
+        : AsyncSeq<int64 * ITimelineEvent<EventBody> array> =
+        let rec loop batchCount pos : AsyncSeq<int64 * ITimelineEvent<EventBody> array> = asyncSeq {
+            match maxPermittedBatchReads with
+            | Some mpbr when batchCount >= mpbr -> log.Information "batch Limit exceeded"; invalidOp "batch Limit exceeded"
+            | _ -> ()
+
+            let batchLog = log |> Log.prop "batchIndex" batchCount
+            let! slice = readSlice pos batchLog
+
+            match slice.Messages with
+            | [||] -> yield -1L, Array.empty
+            | _ ->
+                let version = slice.LastVersion
+                yield version, slice.Messages
+                if not slice.IsEnd then
+                    yield! loop (batchCount + 1) (version + 1L) }
+        loop 0 startPosition
+
+    let resolvedEventBytes events = events |> Array.sumBy resolvedEventLen
+    let logBatchRead streamName t events (batchSize: int64) version (log : ILogger) =
+        let bytes, count = resolvedEventBytes events, events.Length
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let batches = (events.Length - 1)/ (int batchSize) + 1
+        let action = "Load"
+        let evt = Log.Metric.Batch (batches, reqMetric)
+        (log |> Log.prop "bytes" bytes |> Log.event evt).Information(
+            "Mdb{action:l} stream={stream} count={count}/{batches} version={version}",
+            action, streamName, count, batches, version)
+
+    let logLastEventRead streamName t events version (log: ILogger) =
+        let bytes = resolvedEventBytes events
+        let count = 1
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let evt = Log.Metric.ReadLast reqMetric
+        let batches = 1
+        (log |> Log.prop "bytes" bytes |> Log.event evt).Information(
+            "Mdb{action:l} stream={stream} count={count}/{batches} version={version}",
```

Comment:
the `/{batches}` bit can go - the `ReadL` implies that (the log entry should be legible and brief; consistency for regexing is a misfeature)

Reply:
Makes, sense. Removed